### PR TITLE
Better octal and hex-entity decode

### PR DESCRIPTION
**Pull request by GreyWyvern**
# Type of pull request

* [x] Bug fix (involves code and configuration changes)
* [ ] New feature (involves code and configuration changes)
* [ ] Documentation update
* [ ] Something else

# About

A follow-up and improvement over #627.

Octal strings can include series of backslashes of arbitrary length. If there is an odd number of backslashes, a following octal code is valid, but if there's an even number, the following octal code should not be translated.

Previously PdfParser would only account for two backslashes directly preceding an octal code. A commit from in-progress PR #634 extended this to three which probably covers 99.99% of all cases. This change ups that to 100% in that there could be a string with any number of backslashes in a row, and codes will be correctly translated.

Also update decodeEntities() to use a preg_replace_callback() instead of the bulkier preg_split() + foreach loop. Make sure it matches all hexadecimal digits including a-f.

Add new tests for both of these.

# Checklist for code / configuration changes

*In case you changed the code/configuration, please read each of the following checkboxes as they contain valuable information:*

* [x] Please add at least **one test case** (unit test, system test, ...) to demonstrate that the change is working. If existing code was changed, your tests cover these code parts as well.
     By the way, you don't have to provide a full fledged PDF file to demonstrate a fix. Instead a unit test may be sufficient sometimes, please have a look at [FontTest](https://github.com/smalot/pdfparser/blob/master/tests/PHPUnit/Unit/FontTest.php#L40) for example code.
     Code changes without any tests are likely to be rejected. If you dont know how to write tests, no problem, tell us upfront and we may add them ourselves or discuss other ways.
* [x] Please run **PHP-CS-Fixer** before committing, to confirm with our coding styles. See https://github.com/smalot/pdfparser/blob/master/.php-cs-fixer.php for more information about our coding styles.

### CODE REVIEW

**Review by k00ni — APPROVED** [review]
Nice and clean. Thank you again. :+1:

File: src/Smalot/PdfParser/Font.php [hunk 1]
```diff
@@ -368,16 +372,9 @@ public static function decodeOctal(string $text): string
...
-                $text .= \chr(hexdec(trim($part, '#')));
-            } else {
-                $text .= $part;
-            }
-        }
+        $text = preg_replace_callback('/#([0-9a-f]{2})/i', function ($m) {
+            return \chr(hexdec($m[1]));
+        }, $text);
 
         return $text;
```
Comment (k00ni) [inline src/Smalot/PdfParser/Font.php:379]:
```suggestion
        return preg_replace_callback('/#([0-9a-f]{2})/i', function ($m) {
            return \chr(hexdec($m[1]));
        }, $text);
```

File: src/Smalot/PdfParser/Font.php [hunk 2]
```diff
@@ -349,16 +349,20 @@ public static function decodeHexadecimal(string $hexa, bool $add_braces = false)
...
+            return \chr(octdec($m[1]));
+        }, $text);
+
+        // Unescape any parentheses
+        $text = str_replace(['\\(', '\\)'], ['(', ')'], $text);
+
+        // Replace instances of the special string with a single backslash
+        $text = str_replace('[**pdfparserdblslsh**]', '\\', $text);
 
         return $text;
```
Comment (k00ni) [inline src/Smalot/PdfParser/Font.php:367]:
```suggestion
        return str_replace('[**pdfparserdblslsh**]', '\\', $text);
```

**Comment by k00ni** [comment]
Ready for merge?

**Comment by GreyWyvern** [comment]
> Ready for merge?

Yes. All done. 👍

**Pull request merged**